base-trezor-keyring.js

```diff
@@ -0,0 +1,562 @@
+    });
+  }
+
+  deserialize(opts = {}) {
+    this.hdPath = opts.hdPath || hdPathString;
+    this.accounts = opts.accounts || [];
+    this.page = opts.page || 0;
+    this.perPage = opts.perPage || 5;
+    return Promise.resolve();
+  }
+
+  isUnlocked() {
+    return Boolean(this.hdk && this.hdk.publicKey);
+  }
+
+  unlock() {
+    if (this.isUnlocked()) {
+      return Promise.resolve('already unlocked');
+    }
+    return new Promise((resolve, reject) => {
+      this._getPublicKey({
+        path: this.hdPath,
+        coin: 'ETH',
+      })
+        .then((response) => {
+          if (response.success) {
+            this.hdk.publicKey = Buffer.from(response.payload.publicKey, 'hex');
+            this.hdk.chainCode = Buffer.from(response.payload.chainCode, 'hex');
+            resolve('just unlocked');
+          } else {
+            reject(
+              new Error(
+                (response.payload && response.payload.error) || 'Unknown error',
+              ),
+            );
+          }
+        })
+        .catch((e) => {
+          reject(new Error((e && e.toString()) || 'Unknown error'));
+        });
+    });
+  }
+
+  setAccountToUnlock(index) {
+    this.unlockedAccount = parseInt(index, 10);
+  }
+
+  addAccounts(n = 1) {
+    return new Promise((resolve, reject) => {
+      this.unlock()
+        .then((_) => {
+          const from = this.unlockedAccount;
+          const to = from + n;
+
+          for (let i = from; i < to; i++) {
+            const address = this._addressFromIndex(pathBase, i);
+            if (!this.accounts.includes(address)) {
+              this.accounts.push(address);
+            }
+            this.page = 0;
+          }
+          resolve(this.accounts);
+        })
+        .catch((e) => {
+          reject(e);
+        });
+    });
+  }
+
+  getFirstPage() {
+    this.page = 0;
+    return this.__getPage(1);
+  }
+
+  getNextPage() {
+    return this.__getPage(1);
+  }
+
+  getPreviousPage() {
+    return this.__getPage(-1);
+  }
+
+  __getPage(increment) {
+    this.page += increment;
+
+    if (this.page <= 0) {
+      this.page = 1;
+    }
+
+    return new Promise((resolve, reject) => {
+      this.unlock()
+        .then((_) => {
+          const from = (this.page - 1) * this.perPage;
+          const to = from + this.perPage;
+
+          const accounts = [];
+
+          for (let i = from; i < to; i++) {
+            const address = this._addressFromIndex(pathBase, i);
+            accounts.push({
+              address,
+              balance: null,
+              index: i,
+            });
+            this.paths[ethUtil.toChecksumAddress(address)] = i;
+          }
+          resolve(accounts);
+        })
+        .catch((e) => {
+          reject(e);
+        });
+    });
+  }
+
+  getAccounts() {
+    return Promise.resolve(this.accounts.slice());
+  }
+
+  removeAccount(address) {
+    if (
+      !this.accounts.map((a) => a.toLowerCase()).includes(address.toLowerCase())
```

Comment:
We don't need to iterate over all accounts every time.
```suggestion
      this.accounts.every((a) => a.toLowerCase() !== address.toLowerCase())
```